Sources/GRPC/AsyncAwaitSupport/GRPCAsyncServerHandler.swift

```diff
@@ -0,0 +1,519 @@
+/*
+ * Copyright 2021, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import _NIOConcurrency
+import Logging
+import NIOCore
+import NIOHPACK
+
+#if compiler(>=5.5)
+
+@available(macOS 12, iOS 15, tvOS 15, watchOS 8, *)
+public final class GRPCAsyncServerHandler<
```

Comment:
> Who creates an `GRPCAsyncServerHandler`?

It's created in in generated code (an example of which is not present in this PR). The generated code will make use of the initializers defined in the extension near the bottom of this file: https://github.com/grpc/grpc-swift/pull/1249/files#diff-fcf632a779a41e9ebc055fb40ac26d96379a72ca8506cc6e43d022834b9b287fR402-R517.

> For which request/response type is this class used? For all? Only unary?

There is one such initializer per RPC type, so this is a common handler for all of them, which was something @glbrntt was keen we try and do with this API (the existing one has different handler API types per RPC type which we wanted to avoid).

Taking the Echo service, the codegen would generate an `Echo_AsyncEchoProvider` protocol that they would need to conform to (provide just the user functions). The generated code would also contain the following extension which glues it all together:

```swift
@available(macOS 12, iOS 15, tvOS 15, watchOS 8, *)
extension Echo_AsyncEchoProvider {
  public var serviceName: Substring { return "echo.Echo" }

  /// Determines, calls and returns the appropriate request handler, depending on the request's method.
  /// Returns nil for methods not handled by this service.
  public func handle(
    method name: Substring,
    context: CallHandlerContext
  ) -> GRPCServerHandlerProtocol? {
    switch name {
    case "Get":
      return AsyncServerHandler(
        context: context,
        requestDeserializer: ProtobufDeserializer<Echo_EchoRequest>(),
        responseSerializer: ProtobufSerializer<Echo_EchoResponse>(),
        interceptors: self.interceptors?.makeGetInterceptors() ?? [],
        wrapping: self.get(request:context:)
      )

    case "Expand":
      return AsyncServerHandler(
        context: context,
        requestDeserializer: ProtobufDeserializer<Echo_EchoRequest>(),
        responseSerializer: ProtobufSerializer<Echo_EchoResponse>(),
        interceptors: self.interceptors?.makeExpandInterceptors() ?? [],
        wrapping: self.expand(request:responseStreamWriter:context:)
      )

    case "Collect":
      return AsyncServerHandler(
        context: context,
        requestDeserializer: ProtobufDeserializer<Echo_EchoRequest>(),
        responseSerializer: ProtobufSerializer<Echo_EchoResponse>(),
        interceptors: self.interceptors?.makeCollectInterceptors() ?? [],
        wrapping: self.collect(requests:context:)
      )

    case "Update":
      return AsyncServerHandler(
        context: context,
        requestDeserializer: ProtobufDeserializer<Echo_EchoRequest>(),
        responseSerializer: ProtobufSerializer<Echo_EchoResponse>(),
        interceptors: self.interceptors?.makeUpdateInterceptors() ?? [],
        wrapping: self.update(requests:responseStreamWriter:context:)
      )

    default:
      return nil
    }
  }
}
```

This is structurally similar to what we have today except that we implement everything in terms of a single handler and use the various `init(...:wrapping:)` initializers in the codegen.